lib/active_record/connection_adapters/cockroachdb_adapter.rb

```diff
@@ -205,6 +205,20 @@ def initialize(connection, logger, conn_params, config)
           version_num = 202
         end
         @crdb_version = version_num
+
+        Thread.new do
+          begin
+            query = "SELECT crdb_internal.increment_feature_counter('ActiveRecord %d.%d')"
```

Comment:
does this use the same connection or a different one? my fear with async threads is that they cause surprises to the main thread such as closing a connection accidentally.

(very sad about the connection param deas)

Reply:
we should probably also gate this with an environment variable, e.g. `BLOCK_COCKROACHDB_ACTIVERECORD_TELEMETRY` which prevents spawning this thread if the user so wishes

Reply:
env var or some other configuration makes sense. i was asking @keithdoggett if we could also have configuration for using follower reads also, so whatever the config is, it should be a similar pattern for both

Reply:
I think something like #192 should work here. It would allow the user to set it in their `databases.yml` file and would be available in the `@config` hash here.

@otan to your point, it would be better to use `pool.with_connection` inside the thread because otherwise it will create zombie connections that may or may not be dealt with later. So it would look something like this

```ruby
Thread.new do
  pool.with_connection do |conn|
    begin
      query = "SELECT crdb_internal.increment_feature_counter('ActiveRecord %d.%d')"
      conn.query_value(query % [ActiveRecord::VERSION::MAJOR, ActiveRecord::VERSION::MINOR])
```

The connection is returned to the pool after the block finishes.

@rafiss I'm seeing multiple issues in that test class as well. I think it might be due to an error being raised in the thread and it's causing a different than expected error.

Reply:
Thanks for all the helpful feedback. I made the following changes:

- Used `pool.with_connection`
- Added check for `conn.active?`
- Added `disable_cockroachdb_telemetry` config variable
- Added another `rescue` clause to catch and log any unexpected exceptions since we never really want these to bubble up and cause more trouble.

Let's see if tests are happier now.

Reply:
Hmm, it looks like the `pool.with_connection` change causes this to be called three times when I run tests rather than just once. Perhaps because calling `with_connection` triggers another connection to be initialized? Any thoughts on how to avoid this @keithdoggett?

Reply:
Hmm, that's interesting. I thought that `with_connection` just checks out a connection from the pool, but maybe since we're calling this in `initialize`, no connections are in the pool yet and it creates one. I'll have some time to look into this more tomorrow.

Reply:
I looked into this a little bit and I think it's going to be hard to ensure that starting a new thread won't create a new connection within the `initialize` method. `with_connection` checks if there's a connection in the pool (on the first run there's not), if not it creates a new connection which ends up in `initialize` again and it tries to get a free connection from the pool or create a new one. So this will end up creating up to `MAX_THREAD_SIZE` connections or run until one becomes free and our new thread can just use that.

Do you want this to run once per connection or once every time we use the adapter? I wrote a little modification to the `ConnectionPool` that will run this in a new thread on initialization and only creates 1 connection, but it's only run when `ConnectionPool` is initialized.

If you want it to be run on every connection, one way we could go about it is to add some flag to the connection that indicates if this was run yet and keep track of the connections to wait until they're free to run them in a new thread. I'm not exactly sure how it would be implemented in a thread-safe manner since none of the thread-safe methods that `ConnectionPool` implements let you specify a connection.

Reply:
Once per use of the adapter would actually be preferable. If you'd like to share your `ConnectionPool` version that would be appreciated!

Reply:
https://github.com/keithdoggett/activerecord-cockroachdb-adapter/commit/a4c4580ec098bdd108cf2410490a7afd77eb0a11#. I ran the tests with this and there were 5 failures, which all seem to be due to the test trying to force only invalid connections to be available, but `with_connection` created a valid connection each time.
 
It might make sense to test this in a similar way to https://github.com/cockroachdb/activerecord-cockroachdb-adapter/commit/5f14dd422649ad9a68f594d3d555f194ed55825f where we change `config.yml`, `build/config.teamcity.yml`, and https://github.com/cockroachdb/activerecord-cockroachdb-adapter/commit/5f14dd422649ad9a68f594d3d555f194ed55825f#diff-8a06c7c920b8fb065a664471bffa5e1bdfca416b8eaea68ee2e0caec0026f8acR20 to set `disable_cockroachdb_telemetry` to `false` and add a new test to the `postgresql_adapter_test.rb` file that allows the thread to run.

Another option would be to establish a new connection before each of those tests that disables telemetry, then reset to the standard connection afterwards.